### PR TITLE
Move ToTransportAddress to TransportInfrastructure

### PR DESCRIPTION
**Pull request by timbussmann**
## Todo

- [x] Add obsolete messages
- [x] QueueAddress ctor
- [x] Acct Test to test all DI access
  - [x] ReceiveAddresses
  - [x] ITransportAddressResolver
  - [x] also check on send-only endpoints!
- [x] Consider add the ReceiveAddress to IMessageReceiver
- [x] Make sure we use QueueAddress everywhere where it makes sense
- [x] Create ReceiveAddresses builder on ReceiveComponent
- [x] QueueAddress toAString unit tests

### CODE REVIEW

File: src/NServiceBus.Core.Tests/ApprovalFiles/APIApprovals.ApproveNServiceBus.approved.txt [hunk 1]
```diff
@@ -806,6 +806,10 @@ namespace NServiceBus
     {
         public static NServiceBus.Outbox.OutboxSettings EnableOutbox(this NServiceBus.EndpointConfiguration config) { }
     }
+    public static class PathChecker
```
[the remark this answers is not in the document]
Comment (andreasohlund) [reply src/NServiceBus.Core.Tests/ApprovalFiles/APIApprovals.ApproveNServiceBus.approved.txt:809]: Lets copy it over?

File: src/NServiceBus.Core/Features/FeatureConfigurationContext.cs [hunk 2]
```diff
@@ -66,10 +66,10 @@ public class FeatureConfigurationContext
         /// <param name="runtimeSettings">Transport runtime settings.</param>
         /// <param name="recoverabilityPolicy">Recoverability policy to be if processing fails.</param>
         /// <param name="onMessage">The message func.</param>
-        public void AddSatelliteReceiver(string name, string transportAddress, PushRuntimeSettings runtimeSettings, Func<RecoverabilityConfig, ErrorContext, RecoverabilityAction> recoverabilityPolicy, OnSatelliteMessage onMessage)
+        public void AddSatelliteReceiver(string name, QueueAddress transportAddress, PushRuntimeSettings runtimeSettings, Func<RecoverabilityConfig, ErrorContext, RecoverabilityAction> recoverabilityPolicy, OnSatelliteMessage onMessage)
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Features/FeatureConfigurationContext.cs:69]: The parameter name `transportAddress` feels off? => `queueAddress`?

File: src/NServiceBus.Core/Pipeline/SatelliteDefinition.cs [hunk 3]
```diff
@@ -16,7 +16,7 @@ public SatelliteDefinition(string name, string receiveAddress, PushRuntimeSettin
 
         public string Name { get; }
 
-        public string ReceiveAddress { get; }
+        public QueueAddress ReceiveAddress { get; }
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Pipeline/SatelliteDefinition.cs:19]: Rename to "QueueAddress" ?

File: src/NServiceBus.Core/Receiving/ReceiveAddresses.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using Features;
+
+    //TODO should this be an interface?
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Receiving/ReceiveAddresses.cs:6]: I think a class is fine here

File: src/NServiceBus.Core/Receiving/ReceiveAddresses.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    /// <summary>
+    /// Lists the transport addresses this endpoint is consuming messages from.
+    /// </summary>
+    public class ReceiveAddresses
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Receiving/ReceiveAddresses.cs:10]: Reading the comment above I wonder if `TransportAddresses` would be a better name?

File: src/NServiceBus.Core/Receiving/ReceiveComponent.Configuration.cs [hunk 6]
```diff
@@ -22,19 +22,20 @@ public static Configuration PrepareConfiguration(HostingComponent.Configuration
             var purgeOnStartup = settings.PurgeOnStartup;
 
             var transportDefinition = transportSeam.TransportDefinition;
-            var localAddress = transportDefinition.ToTransportAddress(new QueueAddress(queueNameBase, null, null, null));
+            var localAddress = transportDefinition.ToTransportAddress(new QueueAddress(queueNameBase));
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Receiving/ReceiveComponent.Configuration.cs:25]: In theory we could get rid of this and always perform this resolution in the extension methods?

File: src/NServiceBus.Core/Receiving/ReceiveComponent.cs [hunk 7]
```diff
@@ -28,6 +28,16 @@ partial class ReceiveComponent
             HostingComponent.Configuration hostingConfiguration,
             PipelineSettings pipelineSettings)
         {
+            hostingConfiguration.Services.AddSingleton(sp =>
+            {
+                var addressResolver = sp.GetRequiredService<ITransportAddressResolver>();
+                return new ReceiveAddresses(
+                    //TODO: should the main address be null on a send-only endpoint?
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Receiving/ReceiveComponent.cs:35]: What would happen if we don't register the ReceiveAddresses at all for send only endpoints?

File: src/NServiceBus.Core/Transports/Learning/PathChecker.cs [hunk 8]
```diff
@@ -3,8 +3,16 @@ namespace NServiceBus
     using System;
     using System.IO;
 
-    static class PathChecker
+
+    /// <summary>
+    /// TODO (should this be moved to a different namespace)
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Transports/Learning/PathChecker.cs:8]: Lets copy paste ?

File: src/NServiceBus.AcceptanceTests/Core/TransportSeam/When_transport_is_not_configured.cs [hunk 9]
```diff
@@ -0,0 +1,25 @@
+﻿namespace NServiceBus.AcceptanceTests.Core.TransportSeam
+{
+    using System;
+    using NServiceBus.Configuration.AdvancedExtensibility;
+    using NUnit.Framework;
+
+    public class When_transport_is_not_configured : NServiceBusAcceptanceTest
```
Comment (timbussmann) [inline src/NServiceBus.AcceptanceTests/Core/TransportSeam/When_transport_is_not_configured.cs:7]: this doesn't necessarily have to be an acceptace tests?

File: src/NServiceBus.AcceptanceTests/Core/TransportSeam/When_transport_is_started.cs [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+                    testContext.ResolvedAddress = resolver.ToTransportAddress(new QueueAddress("SomeAddress"));
+                    testContext.ReceiveAddresses = receiveAddresses;
+                    return Task.CompletedTask;
+                }
+
+                protected override Task OnStop(IMessageSession session, CancellationToken cancellationToken = default)
+                {
+                    return Task.CompletedTask;
+                }
+
```
Comment (timbussmann) [inline src/NServiceBus.AcceptanceTests/Core/TransportSeam/When_transport_is_started.cs:107]:
```suggestion
```

File: src/NServiceBus.AcceptanceTests/Core/TransportSeam/When_transport_is_started_send_only.cs [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    {
+        [Test]
+        public async Task Should_provide_access_to_address_resolution()
+        {
+            var context = await Scenario.Define<Context>()
+                .WithEndpoint<Endpoint>()
+                .Done(c => c.EndpointsStarted)
+                .Run();
+
+
```
Comment (timbussmann) [inline src/NServiceBus.AcceptanceTests/Core/TransportSeam/When_transport_is_started_send_only.cs:22]:
```suggestion
```

File: src/NServiceBus.AcceptanceTests/NServiceBus.AcceptanceTests.csproj [hunk 12]
```diff
@@ -33,6 +33,9 @@
 
   <ItemGroup>
     <RemoveSourceFileFromPackage Include="Core\**\*.cs" />
+    <RemoveSourceFileFromPackage Remove="Core\TransportSeam\When_transport_is_not_configured.cs" />
+    <RemoveSourceFileFromPackage Remove="Core\TransportSeam\When_transport_is_started.cs" />
+    <RemoveSourceFileFromPackage Remove="Core\TransportSeam\When_transport_is_started_send_only.cs" />
```
Comment (timbussmann) [inline src/NServiceBus.AcceptanceTests/NServiceBus.AcceptanceTests.csproj:38]:
```suggestion
```

I assume this wasn't added intentionally @andreasohlund ?

File: src/NServiceBus.Core/SettingsExtensions.cs [hunk 13]
```diff
@@ -30,16 +31,30 @@ public static string EndpointName(this IReadOnlySettings settings)
...
         /// </summary>
+        [ObsoleteEx(
+            Message = "Inject the ReceiveAddresses class to access the endpoint's receiving transport addresses at runtime. See the NServiceBus version 8 upgrade guide for further details.",
+            TreatAsErrorFromVersion = "9",
+            RemoveInVersion = "10")]
         public static string LocalAddress(this IReadOnlySettings settings)
         {
             Guard.AgainstNull(nameof(settings), settings);
 
+            if (!settings.TryGet<TransportDefinition>(out var transportDefinition))
```
Comment (timbussmann) [inline src/NServiceBus.Core/SettingsExtensions.cs:42]: do we need this check at all? It isn't possible to configure an endpoint without transport IIRC and you can't access LocalAddress before the endpoint is configured/started which is done by the next check?

File: src/NServiceBus.Core/SettingsExtensions.cs [hunk 14]
```diff
@@ -30,16 +31,30 @@ public static string EndpointName(this IReadOnlySettings settings)
...
+                throw new InvalidOperationException("LocalAddress isn't available until the transport is configured.");
+            }
+
             if (!settings.TryGet<ReceiveComponent.Configuration>(out var receiveConfiguration))
             {
-                throw new InvalidOperationException("LocalAddress isn't available since this endpoint is configured to run in send-only mode.");
+                throw new InvalidOperationException("LocalAddress isn't available until the endpoint configuration is complete.");
+            }
+
+            if (receiveConfiguration.IsSendOnlyEndpoint)
```
Comment (timbussmann) [inline src/NServiceBus.Core/SettingsExtensions.cs:52]: should that check go first?

File: src/NServiceBus.Core/Transports/IMessageReceiver.cs [hunk 15]
```diff
@@ -32,6 +32,11 @@ public interface IMessageReceiver
         /// The unique identifier of this instance.
         /// </summary>
         string Id { get; }
+
+        /// <summary>
+        /// The transport address this received is receiving messages from.
```
Comment (timbussmann) [inline src/NServiceBus.Core/Transports/IMessageReceiver.cs:37]:
```suggestion
        /// The transport address this receiver is consuming messages from.
```

**Review by andreasohlund — APPROVED** [review]

**Pull request merged**